Stop trying to load images on an incompatible OS

johnstep opened the pull request:
Signed-off-by: John Stephens <johnstep@docker.com>

This improves the error message when attempting to load an image on an incompatible OS.

Before:
```
Windows:
$ docker load -i alpine.tar
3fb66f713c9f: Loading layer [==================================================>]  4.221MB/4.221MB
re-exec error: exit status 1: output: ProcessBaseLayer C:\ProgramData\Docker\windowsfilter\0bd532c7245dab3d348403edae359732fa64bb2d5bd3ffad62e0240853db5477: The system cannot find the path specified.

Linux:
$ docker load -i nanoserver.tar
open /var/lib/docker/tmp/docker-import-911583298/932e9fb6732a2d14dee02c92a450e6adb13a9f14488a74e71a0fe3b48d16cab9\layer.tar: no such file or directory
```

After:
```
Windows:
$ docker load -i alpine.tar
cannot load linux image on windows

Linux:
$ docker load -i nanoserver.tar
cannot load windows image on linux
```

@dhiltgen @friism @jhowardmsft

johnstep committed:
Stop trying to load images on an incompatible OS

Signed-off-by: John Stephens <johnstep@docker.com>

tonistiigi commented on image/tarexport/load.go:
```diff
@@ -77,6 +78,9 @@ func (l *tarexporter) Load(inTar io.ReadCloser, outStream io.Writer, quiet bool)
 			return err
 		}
+		if img.OS != runtime.GOOS {
+			return fmt.Errorf("cannot load %s image on %s", img.OS, runtime.GOOS)
```
nit: `errors.Errorf`

johnstep replied:
Makes sense, I will take a look, and also test the legacy code path.

thaJeztah replied:
Looks like the check should be less strict, and also accept `img.OS == ""`, or just change this to a Warning.

Current CI is failing because of this, because the `empty` image apparently doesn't have an `OS`;

```
02:30:05 ---> Making bundle: .ensure-emptyfs (in bundles/17.06.0-dev/test-integration-cli)
02:30:05 ++++ tar -cC bundles/17.06.0-dev/test-integration-cli/emptyfs .
02:30:05 ++++ docker load
02:30:05 cannot load  image on linux
```

Could there be other cases where `OS` doesn't have to match actual OS? (idk, things could be emulated inside the container or VM)

johnstep replied:
Thanks! Yeah, I actually thought of that case, but had not addressed it yet. I'm not sure about other cases, but will look into it.

friism replied:
@thaJeztah I don't see the point of making it a warning. The load will fail hard after the warning is issued.

tonistiigi reviewed (commented):
I'm wondering where that error comes in linux with the current version and what ID is `932e9fb6732a2d14dee02c92a450e6adb13a9f14488a74e71a0fe3b48d16cab9`. Would like to confirm that the tarballs saved in windows actually comply to the spec.

friism commented:
@tonistiigi 

> I'm wondering where that error comes in linux with the current version

I just tested, trying to load windows image into linux daemon. It's a hard failure. cc @thaJeztah 

```
PS C:\Users\friism> docker load -i .\nano.tar
open /var/lib/docker/tmp/docker-import-783618241/932e9fb6732a2d14dee02c92a450e6adb13a9f14488a74e71a0fe3b48d16cab9\layer.tar: no such file or directory
```

tonistiigi commented:
@friism It is probably just the backslash in the path.

friism commented:
@tonistiigi what are you trying to get at? That it should be a supported use-case to handle Windows images with a linux daemon?

thaJeztah commented:
@friism the discussion was that any image (windows, linux, foofoobar) uses the same format, so import/export, save/load (and even push/pull) should be possible, without the engine making assumptions about the image. The OS field "should" be set, but _can_ be empty (what we've actually seen a while back in the official Microsoft images); so the better option would probably be to proceed with the load action, but output a warning that the image may not be supported on the platform.

friism commented:
@thaJeztah I don't understand the details, but I'm not sure that an NTFS-filesystem Windows image can exist unpacked in a Linux graphdriver and I'm even more sure that the reverse (Linux image in Windows graphdriver) cannot happen without loosing file metadata, etc.

jhowardmsft commented:
LGTM

johnstep commented:
I updated this to only fail when there is a mismatch between Windows and some other OS, and also added a special case for Linux Containers on Windows. This should address all the previous concerns, while still achieving the desired result.

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
Janky is hitting https://github.com/moby/moby/issues/33863
